### PR TITLE
Implements a CleanPendingUploadsJob, a FindPendingUploadsFailures query, and the Rake task clean:pending_uploads

### PR DESCRIPTION
**Pull request by jrgriffiniii**
These provide one with the ability to delete resources which have failed file uploads as members. Introduced in response to #3488

### CODE REVIEW

**Comment by jrgriffiniii** [comment]
> Can we add a check so that we don't delete anything that has filesets? All the ones we want to delete have pending up loads but no file sets; I'm not sure what else may be out there with stale pending uploads.

I have added an additional test for this at https://github.com/pulibrary/figgy/pull/3493/files#diff-242bcc50b5849636074e348c21bbb49bR56

**Comment by jrgriffiniii** [comment]
> And maybe a dry-run flag that just returns them so the task can list them by title and date modified?

This was added with https://github.com/pulibrary/figgy/pull/3493/files#diff-242bcc50b5849636074e348c21bbb49bR73

File: app/jobs/clean_pending_uploads_job.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+class CleanPendingUploadsJob < ApplicationJob
+  def perform(dry_run: false)
+    query_service.custom_queries.find_pending_upload_failures.each do |resource|
+      change_set_persister.buffer_into_index do |buffered_changeset_persister|
+        if resource.respond_to?(:pending_uploads) && resource.pending_uploads.empty? && resource.state.include?("pending")
+          change_set = DynamicChangeSet.new(resource)
+          if dry_run
+            Valkyrie.logger.info("Found #{resource.id} as an uploaded resource without any FileSets - this would normally be deleted by CleanPendingUploadsJob")
```
Comment (hackmastera) [inline app/jobs/clean_pending_uploads_job.rb]: Can we also print the decorated title?

File: app/jobs/clean_pending_uploads_job.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+          else
+            buffered_changeset_persister.delete(change_set: change_set)
+          end
+          Valkyrie.logger.info "Deleted a resource with failed uploads with the ID: #{resource.id}"
```
Comment (hackmastera) [inline app/jobs/clean_pending_uploads_job.rb]: This should be inside the else

File: lib/tasks/clean.rake [hunk 3]
```diff
@@ -16,4 +16,10 @@ namespace :clean do
   task :dead_queues do
     CleanDeadQueuesJob.set(queue: :low).perform_later
   end
+
+  desc "Clean failed PendingUpload Resources."
+  task :pending_uploads, [:dry_run] => [:environment] do |_t, args|
+    args.with_defaults(dry_run: false)
+    CleanPendingUploadsJob.set(queue: :low).perform_later(dry_run: args[:dry_run].present?)
```
Comment (hackmastera) [inline lib/tasks/clean.rake]: Perhaps the job should take a logger and dry-run should do perform_now with a stdout logger.

File: app/queries/find_pending_upload_failures.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  attr_reader :query_service
+  delegate :resources, to: :query_service
+  delegate :resource_factory, to: :query_service
+  def initialize(query_service:)
+    @query_service = query_service
+  end
+
+  def find_pending_upload_failures
+    relation = { "pending_uploads" => [], "state" => ["pending"] }
```
Comment (jrgriffiniii) [inline app/queries/find_pending_upload_failures.rb]: This should be querying for cases where the `file` uploads have been removed (https://github.com/pulibrary/figgy/blob/master/app/services/file_appender.rb#L175), but the FileSet itself has not had the necessary `FileMetadata` nodes added.

**Comment by jrgriffiniii** [comment]
https://github.com/pulibrary/figgy/blob/master/app/services/file_appender.rb#L5 and https://github.com/pulibrary/figgy/blob/master/app/services/file_appender.rb#L8 should be raised within `FileAppender`, but I'm not certain where it is.

File: app/queries/find_pending_upload_failures.rb (hunk 4 above)
Comment (jrgriffiniii) [reply app/queries/find_pending_upload_failures.rb]: https://app.honeybadger.io/projects/53391/faults/57256942 shows that https://github.com/pulibrary/figgy/blob/master/app/models/pending_upload.rb#L13 raises an error, but https://github.com/pulibrary/figgy/blob/master/app/services/file_appender.rb#L178 should remove `PendingUploads` given that https://github.com/pulibrary/figgy/blob/master/app/services/file_appender.rb#L112 captures and logs errors. This query should probably just query for Resources and update the `pending_uploads` attribute when they are missing the `file_name` attribute.

Comment (jrgriffiniii) [reply app/queries/find_pending_upload_failures.rb]: `PendingUpload` seems to have been persisted with a random UUID for some years (https://github.com/pulibrary/figgy/blame/dd40b22a78e011b18d942364297d271eca1f605b/app/controllers/base_resource_controller.rb#L34), so I will proceed with trying to just query for cases where these have been persisted with invalid metadata.

**Comment by jrgriffiniii** [comment]
https://github.com/pulibrary/figgy/pull/3493/commits/951909cdd93023238586f4f770c0c518c7dabb38 and https://github.com/pulibrary/figgy/pull/3493/commits/2cd14d927dd2d11d9d58b22fcbbe574d9f77626c ensure that `pending_uploads` on Resources are just deleted, while the Resources themselves remain persisted. `pending_uploads` will *not* be deleted for cases where the `file_name` on the `PendingUpload` is not `nil`.

**Comment by tpendragon** [comment]
We've redone how pending uploads work since this PR, and the issue it was solving hasn't cropped up again. Closing.

**Comment by tpendragon** [comment]
@jrgriffiniii @hackartisan Anyone remember why I reopened this?

**Comment by hackartisan** [comment]
There are a bunch of messed up resources. The issue is still valid. Last comment: https://github.com/pulibrary/figgy/issues/3488#issuecomment-648346651 -- we can talk about this on a Monday if you like.

**Comment by bess** [comment]
Looked at this in team meeting -- look again in August 2021

**Pull request closed without merging**